File: src/cli/shared.ts
import { Command, InvalidArgumentError, Option } from "commander";

import { AdapterError } from "../adapters/index.js";
import type { TargetSpec } from "../compare.js";
import type { CliFlags, ResolvedConfig } from "../config.js";
import { assertNever, GymratError, hintOf, messageOf } from "../errors.js";
import { NotAGitRepositoryError, tryGit } from "../git.js";
import { formatHintLabel, formatLabel, highlightInlineCode } from "../report/format.js";
import type { FailOnCondition, ReportOptions } from "../report/types.js";
import type { RunOptions } from "../sampling.js";
import { acquireLock, type ReleaseLock } from "../session/lock.js";
import { lockfilePath, repoRoot } from "../session/paths.js";
import { installTerminationCleanup } from "../signals.js";
import { MAX_TIMEOUT_SECONDS } from "../timer-limits.js";

// ---------------------------------------------------------------------------
// URL constants
// ---------------------------------------------------------------------------

const BUGS_URL = "https://github.com/jeffzi/gymrat/issues";

// ---------------------------------------------------------------------------
// Exit codes
// ---------------------------------------------------------------------------

/** The exit code of a gate trip: a run that did what it was asked and said no. */
export const GATE_EXIT_CODE = 1;

/** The exit code of a tool failure, the convention every unhandled error exits on. */
export const TOOL_FAILURE_EXIT_CODE = 2;

// ---------------------------------------------------------------------------
// Debug mode
// ---------------------------------------------------------------------------

/** Whether the global `--debug` flag was passed on the command line. */
let debugMode = false;

/** Set the global debug flag read by {@link debugMode} to control verbose output. */
export function setDebugMode(value: boolean): void {
  debugMode = value;
}

// ---------------------------------------------------------------------------
// Stream helpers
// ---------------------------------------------------------------------------

/**
 * `@types/node` declares `isTTY` as `boolean`, but Node leaves it `undefined`
 * on non-TTY streams — the comparison is what makes the boolean honest.
 */
export function isTTY(stream: { isTTY?: boolean }): boolean {
  return stream.isTTY === true;
}

/**
 * Write to a stream and resolve once the chunk has been handed on.
 *
 * `process.exit` drops whatever is still queued, so an immediate exit truncates
 * a report the stream has taken but not yet written out. A `true` return does
 * not mean it has — only that the chunk fit under the high-water mark — so
 * completion is taken from the callback `write` reports it on.
 *
 * The `error` listener stays attached on the failing path: the stream emits
 * `error` right after handing the same error to the callback, and with nothing
 * listening that lands as an unhandled `error` event, which tears the process
 * down before the rejection can be acted on.
 */
export function writeAndFlush(stream: NodeJS.WriteStream, data: string): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    const onError = (error: Error): void => {
      reject(error);
    };
    stream.once("error", onError);

    stream.write(data, (error) => {
      if (error) {
        reject(error);
        return;
      }
      stream.removeListener("error", onError);
      resolve();
    });
  });
}

// ---------------------------------------------------------------------------
// Color control
// ---------------------------------------------------------------------------

/**
 * Veto color unconditionally: `styleText` resolves `FORCE_COLOR` before `NO_COLOR`, so a
 * `FORCE_COLOR` left over from the caller's shell would otherwise defeat `NO_COLOR` and leak
 * ANSI escapes into non-interactive output.
 */
export function suppressColor(): void {
  delete process.env.FORCE_COLOR;
  process.env.NO_COLOR = "1";
}

/**
 * Translate Commander's `--no-color` flag into the color override the report
 * renderer reads: `false` vetoes color, `undefined` defers to auto-detection.
 */
export function colorOverrideOf(options: { color: boolean }): false | undefined {
  return options.color ? undefined : false;
}

// ---------------------------------------------------------------------------
// Error formatting
// ---------------------------------------------------------------------------

/**
 * Render an error for stderr with a red `Error:` label, the message body,
 * and optional diagnostic sections in order:
 *
 * 1. **Error: label** — always present, styled red via `formatLabel`.
 * 2. **Body** — `AdapterError` keeps its class-name prefix after the label
 *    (`Error: AdapterError: <message>`); everything else is the bare message.
 * 3. **Stack trace** — included when `options.debug` is true and the error
 *    carries a stack.
 * 4. **Hint** — appended for `GymratError` instances that carry a hint.
 * 5. **Bug-report footer** — appended for errors that are NOT `GymratError`
 *    (unexpected errors the user should report).
 *
 * Color is governed by `styleText` auto-detection: `NO_COLOR` / `FORCE_COLOR`
 * env vars and the stream's TTY status.
 */
export function formatCliError(error: unknown, options: { debug?: boolean } = {}): string {
  const errorLabel = `${formatLabel("Error", "red", process.stderr)}: `;

  let body: string;
  if (error instanceof AdapterError) {
    body = `${error.name}: ${error.message}`;
  } else {
    body = messageOf(error);
  }

  let output = `${errorLabel}${body}`;

  if (options.debug === true && error instanceof Error && error.stack !== undefined) {
    output += `\n${error.stack}`;
  }

  const hint = hintOf(error);
  if (hint !== undefined) {
    output += `\n${formatHintLabel(process.stderr)} ${highlightInlineCode(hint, process.stderr)}`;
  }

  if (!(error instanceof GymratError)) {
    output += highlightInlineCode(
      `\nRun with \`gymrat --debug\` for details. If this is a bug, please report it at\n${BUGS_URL}`,
      process.stderr,
    );
  }

  return output;
}

/** Print a formatted error to stderr and exit on `code`. */
export async function exitWithError(error: unknown, code = TOOL_FAILURE_EXIT_CODE): Promise<never> {
  try {
    await writeAndFlush(process.stderr, `${formatCliError(error, { debug: debugMode })}\n`);
  } catch {
    /*
     * stderr is the reporting channel, so a write failure (closed pipe, EPIPE)
     * has nowhere to be reported. Swallow it rather than let it escape as an
     * unhandled rejection, which would exit 1 — the code reserved for a gate
     * trip — instead of the code this exit was asked for.
     */
  }
  process.exit(code);
}

// ---------------------------------------------------------------------------
// Flag parsers
// ---------------------------------------------------------------------------

/**
 * Parse the `label=target` syntax of a positional argument, baseline or candidate.
 *
 * Only the first `=` splits, so a target containing its own `=` survives intact —
 * `a=b=c` parses to label `a`, target `b=c`.
 */
export function parsePositional(positional: string): TargetSpec {
  const eqIndex = positional.indexOf("=");
  const label = eqIndex === -1 ? undefined : positional.slice(0, eqIndex);
  const target = eqIndex === -1 ? positional : positional.slice(eqIndex + 1);

  /*
   * An empty half is always a typo, and both halves fail silently rather than
   * loudly: an empty target resolves to the working directory, so `old=` would
   * benchmark whatever the user happens to have checked out, and an empty label
   * leaves the report's header cells holding nothing but ANSI escapes.
   */
  if (label === "") {
    throw new InvalidArgumentError(
      'the label before "=" is empty; write the positional as "label=<ref|dir>" or drop the "=".',
    );
  }
  if (target === "") {
    throw new InvalidArgumentError(
      'the target is empty; write the positional as "[label=]<ref|dir>".',
    );
  }

  return { ...(label !== undefined && { label }), target };
}

/** Accumulate parsed candidate positionals as Commander walks the variadic argument. */
export function collectPositional(value: string, previous: readonly TargetSpec[]): TargetSpec[] {
  return [...previous, parsePositional(value)];
}

/**
 * Build a coercer for a numeric flag value, rejecting anything that is not a
 * positive integer at or below `max`.
 */
export function parsePositiveIntegerUpTo(max: number): (value: string) => number {
  return (value: string): number => {
    const parsed = Number(value);
    if (!/^\d+$/.test(value) || parsed <= 0) {
      throw new InvalidArgumentError("must be a positive integer.");
    }
    if (parsed > max) {
      throw new InvalidArgumentError(`must be a positive integer no greater than ${max}.`);
    }
    return parsed;
  };
}

/** Accept a complete finite decimal — reject trailing garbage, Infinity, and NaN. */
export function parseStopTargetValue(value: string): number {
  if (!/^[+-]?(\d+\.?\d*|\.\d+)([eE][+-]?\d+)?$/.test(value)) {
    throw new InvalidArgumentError("must be a finite number.");
  }
  return Number(value);
}

/** Parse a strictly positive finite decimal — rejects negatives, zero, Infinity, NaN, and trailing garbage. */
export function parsePositiveNumber(value: string): number {
  if (!/^\d+(\.\d+)?$/.test(value)) {
    throw new InvalidArgumentError("must be a positive number.");
  }
  const parsed = Number(value);
  if (!Number.isFinite(parsed) || parsed <= 0) {
    throw new InvalidArgumentError("must be a positive number.");
  }
  return parsed;
}

/** `parsePositiveNumber` with an upper bound derived from the 32-bit timer ceiling. */
export function parseMaxMinutes(value: string): number {
  const parsed = parsePositiveNumber(value);
  const maxMinutes = Math.floor(MAX_TIMEOUT_SECONDS / 60);
  if (parsed > maxMinutes) {
    throw new InvalidArgumentError(`must be at most ${String(maxMinutes)} minutes.`);
  }
  return parsed;
}

/**
 * Matches the percentage of a `geomean:<pct>` condition.
 */
const GEOMEAN_CONDITION_RE = /^geomean:(-?\d+(?:\.\d+)?)$/;

/**
 * Accepts `regressed` or `geomean:<number>`. Throws `InvalidArgumentError`
 * for anything else so Commander renders the allowed grammar in the usage error.
 */
export function parseFailOn(
  value: string,
  previous: readonly FailOnCondition[],
): FailOnCondition[] {
  if (value === "regressed") {
    return [...previous, { kind: "regressed" }];
  }

  const match = GEOMEAN_CONDITION_RE.exec(value);
  const pct = match ? Number(match[1]) : NaN;
  if (Number.isFinite(pct)) {
    return [...previous, { kind: "geomean", pct }];
  }

  throw new InvalidArgumentError(
    'allowed values are "regressed" or "geomean:<number>" (e.g. geomean:2).',
  );
}

// ---------------------------------------------------------------------------
// Option builders
// ---------------------------------------------------------------------------

/**
 * Declare the flags `resolveConfig` reads, in one place.
 */
export function addConfigOptions(command: Command): Command {
  return command
    .option("-b, --bench <cmd>", "bench command")
    .option("-p, --prepare <script>", "preparation script to run before each revision")
    .option("-a, --adapter <type>", "adapter type for parsing benchmark output")
    .option(
      "-s, --samples <number>",
      "paired samples per target",
      parsePositiveIntegerUpTo(Number.MAX_SAFE_INTEGER),
    )
    .option(
      "-t, --timeout <number>",
      "timeout in seconds",
      parsePositiveIntegerUpTo(MAX_TIMEOUT_SECONDS),
    )
    .option("-c, --config <file>", "configuration file path");
}

/**
 * Declare the flags a benchmarking command carries: the configuration set plus
 * the two that govern how its report prints.
 */
export function addSharedOptions(command: Command): Command {
  return addConfigOptions(command)
    .option("--no-color", "print the report without ANSI styles")
    .addOption(
      new Option("--format <value>", "output format").choices(["text", "json"]).default("text"),
    );
}

// ---------------------------------------------------------------------------
// Flag types
// ---------------------------------------------------------------------------

import type { StrictOmit } from "../errors.js";

/** The flags every command carries: everything `resolveConfig` reads, plus how to print. */
export interface SharedFlags extends CliFlags {
  color: boolean;
  format: "text" | "json";
}

/** The status command's flags: the shared set minus the format choice status does not offer. */
export type StatusFlags = StrictOmit<SharedFlags, "format">;

/** The keep command's flags: the configuration set plus the message the commit carries. */
export interface KeepFlags extends CliFlags {
  message?: string;
}

/**
 * The finalize command's flags — the only loop command whose flags are all its
 * own, because closing a session reads no configuration.
 */
export interface FinalizeFlags {
  message?: string;
  branch?: string;
}

/** The measure command's flags: the shared set plus whether the run becomes history. */
export interface MeasureFlags extends SharedFlags {
  record: boolean;
}

/** The compare command's flags: the shared set plus the two only a verdict can answer. */
export interface CompareFlags extends SharedFlags {
  failOn: FailOnCondition[];
  verbose: boolean;
}

/**
 * The doctor command's flags: config options plus format, color, and the bench skip.
 */
export interface DoctorFlags extends StrictOmit<SharedFlags, "bench"> {
  bench: string | boolean | undefined;
}

/** The init command's flags: wizard inputs, not config overrides. */
export interface InitFlags {
  bench?: string;
  adapter?: string;
  checks?: string;
  stopTarget?: number;
  stopMaxIterations?: number;
  primary?: string;
  runbook?: string | boolean;
  skill?: boolean;
  yes: boolean;
}

/** The supervise command's flags: session caps plus where to write the event log. */
export interface SuperviseFlags {
  maxMinutes: number;
  maxUsd?: number;
  log?: string;
  model?: string;
  allowDirty: boolean;
}

// ---------------------------------------------------------------------------
// Locking
// ---------------------------------------------------------------------------

/**
 * The repository a run must serialize against, or `undefined` when there is none.
 */
function lockableRepoRoot(): string | undefined {
  try {
    return repoRoot();
  } catch (error) {
    if (error instanceof NotAGitRepositoryError) {
      return undefined;
    }
    throw error;
  }
}

/**
 * Run `execute`, and route a thrown error through {@link exitWithError}.
 */
export async function runOrExit<T>(
  execute: () => Promise<T>,
  exitCodeOf: (error: unknown) => number = () => TOOL_FAILURE_EXIT_CODE,
): Promise<T> {
  try {
    return await execute();
  } catch (error) {
    return exitWithError(error, exitCodeOf(error));
  }
}

/**
 * Hold the repository's single-flight lock for the length of `run`, routing a
 * thrown error through {@link exitWithError}.
 */
export async function withRepoLock<T>(
  command: string,
  run: () => Promise<T>,
  exitCodeOf: (error: unknown) => number = () => TOOL_FAILURE_EXIT_CODE,
): Promise<T> {
  let root: string | undefined;
  try {
    root = lockableRepoRoot();
  } catch (error) {
    return exitWithError(error);
  }
  if (root === undefined) {
    return await runOrExit(run, exitCodeOf);
  }

  let release: ReleaseLock;
  try {
    release = acquireLock(lockfilePath(root), command);
  } catch (error) {
    return exitWithError(error);
  }

  process.once("exit", release);

  try {
    return await runOrExit(run, exitCodeOf);
  } finally {
    process.removeListener("exit", release);
    release();
  }
}

// ---------------------------------------------------------------------------
// Run infrastructure
// ---------------------------------------------------------------------------

/** Wire `config`'s run settings and `progress`'s callbacks into the shared run fields. */
export function runOptionsOf(
  config: ResolvedConfig,
  progress: ProgressReporter,
): RunOptions & Required<Pick<RunOptions, "onProgress" | "warn">> {
  return {
    bench: config.bench,
    prepare: config.prepare,
    adapter: config.adapter,
    samples: config.samples,
    timeoutSeconds: config.timeoutSeconds,
    configMetrics: config.metrics,
    configKinds: config.kinds,
    onProgress: (step) => {
      progress.emit(step);
    },
    warn: (message) => {
      progress.warn(message);
    },
  };
}

/**
 * Suppress color per `--no-color`, then build the reporter the run streams its
 * progress through for `targetCount` targets.
 */
export function beginRun(options: SharedFlags, targetCount: number): ProgressReporter {
  if (!options.color) {
    suppressColor();
  }
  const tty = isTTY(process.stderr);
  const colorSuppressed = process.env.NO_COLOR !== undefined;
  return createProgressReporter(tty && !colorSuppressed, tty, targetCount);
}

/**
 * Run `execute`, guarded by `progress`: stop the reporter once the run settles.
 */
export async function runGuarded<T>(
  progress: ProgressReporter,
  execute: () => Promise<T>,
): Promise<T> {
  try {
    return await execute();
  } finally {
    progress.stop();
  }
}

/**
 * Run `execute` with an abort signal a termination signal trips, then exit with
 * the conventional `128 + signum`.
 */
export async function runInterruptibly<T>(
  execute: (signal: AbortSignal) => Promise<T>,
): Promise<T> {
  const run = new AbortController();
  const uninstallTerminationCleanup = installTerminationCleanup(() => {
    run.abort();
  });

  try {
    return await execute(run.signal);
  } finally {
    uninstallTerminationCleanup();
  }
}

/**
 * Render `result` per `options.format` and write it to stdout.
 */
export async function emitReport<T>(
  result: T,
  options: SharedFlags,
  render: { json: (result: T) => string; text: (result: T, textOptions: ReportOptions) => string },
  textOptions: ReportOptions,
): Promise<void> {
  let output: string;
  switch (options.format) {
    case "json":
      output = render.json(result);
      break;
    case "text":
      output = render.text(result, textOptions);
      break;
    default:
      assertNever(options.format);
  }

  await writeAndFlush(process.stdout, output + "\n");
}

// ---------------------------------------------------------------------------
// Git environment
// ---------------------------------------------------------------------------

export interface GitEnvironment {
  gitAvailable: boolean;
  insideGitRepo: boolean;
  repoRootDir?: string;
  gitError?: string;
}

/** Probe git's availability and repository status from `cwd`, without throwing. */
export function detectGitEnvironment(cwd: string): GitEnvironment {
  const gitAvailable = tryGit(["--version"], cwd) === undefined;
  if (!gitAvailable) {
    return { gitAvailable: false, insideGitRepo: false };
  }

  try {
    return { gitAvailable: true, insideGitRepo: true, repoRootDir: repoRoot(cwd) };
  } catch (error) {
    if (error instanceof NotAGitRepositoryError) {
      return { gitAvailable: true, insideGitRepo: false };
    }
    const message = error instanceof Error ? error.message : String(error);
    return { gitAvailable: true, insideGitRepo: true, gitError: message };
  }
}

import { createProgressReporter, type ProgressReporter } from "./progress.js";
export { createProgressReporter, type ProgressReporter } from "./progress.js";

File: src/gymrat_py/cli/shared.py
"""Shared CLI infrastructure: parsing, exit routing, error rendering, locking.

This module holds the pieces every benchmarking command reuses — the argument
coercers, the error formatter and exit path, the repository single-flight lock,
and the render-mode resolution — with no dependency on the heavy statistics
stack or the command bodies, so importing it stays cheap.
"""

import contextlib
import os
import re
import sys
import traceback
from collections.abc import Awaitable, Callable, Sequence
from dataclasses import dataclass
from enum import StrEnum
from typing import Annotated, Literal, NoReturn, Protocol

import typer
from rich.markup import escape

from gymrat_py.adapters.types import AdapterError
from gymrat_py.cli.progress import ProgressReporter, create_progress_reporter
from gymrat_py.cli.status_line import RenderMode
from gymrat_py.config import MAX_TIMEOUT_SECONDS, CliFlags, ResolvedConfig
from gymrat_py.errors import GymratError, hint_of, message_of
from gymrat_py.git import NotAGitRepositoryError
from gymrat_py.report.style import (
    RENDER_WIDTH,
    format_hint_label,
    highlight_inline_code,
    markup,
    render_lines,
)
from gymrat_py.report.types import FailOnCondition, GeomeanFailOn, RegressedFailOn, ReportOptions
from gymrat_py.sampling import RunOptions, TargetSpec
from gymrat_py.session.lock import acquire_lock
from gymrat_py.session.paths import lockfile_path, repo_root

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

BUGS_URL = "https://github.com/jeffzi/gymrat/issues"

# The exit code of a gate trip: a run that did what it was asked and said no.
GATE_EXIT_CODE = 1

# The exit code of a tool failure, the convention every unhandled error exits on.
TOOL_FAILURE_EXIT_CODE = 2

_POSITIVE_INTEGER_RE = re.compile(r"\d+")
_POSITIVE_NUMBER_RE = re.compile(r"\d+(?:\.\d+)?")
_GEOMEAN_CONDITION_RE = re.compile(r"geomean:(-?\d+(?:\.\d+)?)")


class _WritableStream(Protocol):
    """A text stream this module writes error and progress output to."""

    def write(self, data: str, /) -> object: ...

    def flush(self) -> object: ...


# ---------------------------------------------------------------------------
# Debug mode
# ---------------------------------------------------------------------------


class _DebugState:
    """Holds the global ``--debug`` flag without reaching for a ``global`` statement."""

    enabled: bool = False


def set_debug_mode(value: bool) -> None:  # noqa: FBT001 -- 1:1 setter for the --debug flag
    """Set the module debug flag that governs stack traces in error output."""
    _DebugState.enabled = value


# ---------------------------------------------------------------------------
# Stream helpers
# ---------------------------------------------------------------------------


def is_tty(stream: object) -> bool:
    """Whether ``stream`` reports itself as an interactive terminal."""
    isatty = getattr(stream, "isatty", None)
    return bool(isatty()) if callable(isatty) else False


def write_and_flush(stream: _WritableStream, data: str) -> None:
    """Write ``data`` to ``stream`` and flush it so an immediate exit cannot truncate it."""
    stream.write(data)
    stream.flush()


# ---------------------------------------------------------------------------
# Color control
# ---------------------------------------------------------------------------


def suppress_color() -> None:
    """Veto color unconditionally by clearing ``FORCE_COLOR`` and setting ``NO_COLOR``.

    The style layer resolves ``FORCE_COLOR`` before ``NO_COLOR``, so a leftover
    ``FORCE_COLOR`` from the caller's shell would otherwise defeat ``--no-color``.
    """
    os.environ.pop("FORCE_COLOR", None)
    os.environ["NO_COLOR"] = "1"


def color_override_of(color: bool) -> Literal[False] | None:  # noqa: FBT001 -- 1:1 map of the --color flag
    """Translate the ``--color`` flag into the renderer's color override.

    ``False`` vetoes color; ``None`` defers to auto-detection.
    """
    return None if color else False


def _resolve_stderr_color() -> bool:
    """Resolve whether stderr error output should carry color.

    Precedence: an explicit ``FORCE_COLOR`` wins over ``NO_COLOR``, which wins
    over the stream's own TTY detection.
    """
    if os.environ.get("FORCE_COLOR"):
        return True
    if "NO_COLOR" in os.environ:
        return False
    return is_tty(sys.stderr)


# ---------------------------------------------------------------------------
# Error formatting
# ---------------------------------------------------------------------------


def format_cli_error(error: object, *, debug: bool = False) -> str:
    """Render ``error`` for stderr with a red ``Error:`` label and optional sections.

    The sections appear in order: the label, the message body (an
    :class:`AdapterError` keeps its class-name prefix), the stack trace when
    ``debug`` is set, a ``Hint:`` for a :class:`GymratError` that carries one,
    and a report-a-bug footer for errors that are not :class:`GymratError`.
    """
    error_label = f"{markup('Error', 'red')}: "

    if isinstance(error, AdapterError):
        body = f"{type(error).__name__}: {message_of(error)}"
    else:
        body = message_of(error)

    doc = f"{error_label}{escape(body)}"

    if debug and isinstance(error, BaseException) and error.__traceback__ is not None:
        stack = "".join(traceback.format_exception(type(error), error, error.__traceback__))
        doc += f"\n{escape(stack.rstrip())}"

    hint = hint_of(error)
    if hint is not None:
        doc += f"\n{format_hint_label()} {highlight_inline_code(hint)}"

    if not isinstance(error, GymratError):
        footer = (
            "\nRun with `gymrat --debug` for details. "
            "If this is a bug, please report it at\n"
            f"{BUGS_URL}"
        )
        doc += highlight_inline_code(footer)

    return render_lines(doc, color=_resolve_stderr_color(), width=RENDER_WIDTH)


def exit_with_error(error: object, code: int = TOOL_FAILURE_EXIT_CODE) -> NoReturn:
    """Print a formatted error to stderr and exit on ``code``, even if the write fails."""
    rendered = f"{format_cli_error(error, debug=_DebugState.enabled)}\n"
    # stderr is the reporting channel, so a write failure (a closed or broken
    # pipe) has nowhere to be reported. Swallow it rather than let it change the
    # exit code this call was asked for.
    with contextlib.suppress(OSError):
        write_and_flush(sys.stderr, rendered)
    with contextlib.suppress(OSError):
        sys.stdout.flush()
    raise typer.Exit(code)


# ---------------------------------------------------------------------------
# Flag parsers
# ---------------------------------------------------------------------------


def parse_positional(positional: str) -> TargetSpec:
    """Parse the ``label=target`` syntax of a positional, splitting on the first ``=`` only.

    A target containing its own ``=`` survives intact — ``a=b=c`` parses to label
    ``a``, target ``b=c``. An empty half is always a typo, so each raises its own
    usage error rather than resolving to a silent default.
    """
    eq_index = positional.find("=")
    if eq_index == -1:
        label: str | None = None
        target = positional
    else:
        label = positional[:eq_index]
        target = positional[eq_index + 1 :]

    if label == "":
        message = (
            'the label before "=" is empty; '
            'write the positional as "label=<ref|dir>" or drop the "=".'
        )
        raise typer.BadParameter(message)
    if target == "":
        message = 'the target is empty; write the positional as "[label=]<ref|dir>".'
        raise typer.BadParameter(message)

    return TargetSpec(label=label, target=target)


def collect_positional(value: str, previous: Sequence[TargetSpec]) -> list[TargetSpec]:
    """Accumulate parsed candidate positionals as the parser walks the variadic argument."""
    return [*previous, parse_positional(value)]


def parse_positive_integer_up_to(max_value: int) -> Callable[[str], int]:
    """Build a coercer accepting only a positive integer at or below ``max_value``."""

    def parse(value: str) -> int:
        if _POSITIVE_INTEGER_RE.fullmatch(value) is None or int(value) <= 0:
            message = "must be a positive integer."
            raise typer.BadParameter(message)
        parsed = int(value)
        if parsed > max_value:
            message = f"must be a positive integer no greater than {max_value}."
            raise typer.BadParameter(message)
        return parsed

    return parse


def parse_positive_number(value: str) -> float:
    """Parse a strictly positive finite decimal, rejecting negatives, zero, and trailing garbage."""
    if _POSITIVE_NUMBER_RE.fullmatch(value) is None:
        message = "must be a positive number."
        raise typer.BadParameter(message)
    parsed = float(value)
    if parsed <= 0:
        message = "must be a positive number."
        raise typer.BadParameter(message)
    return parsed


def parse_max_minutes(value: str) -> float:
    """Parse a positive number of minutes bounded by the 32-bit timer ceiling."""
    parsed = parse_positive_number(value)
    max_minutes = MAX_TIMEOUT_SECONDS // 60
    if parsed > max_minutes:
        message = f"must be at most {max_minutes} minutes."
        raise typer.BadParameter(message)
    return parsed


def parse_fail_on(value: str) -> FailOnCondition:
    """Parse a fail-on condition: ``regressed`` or ``geomean:<number>``.

    Anything else raises a usage error naming the allowed grammar.
    """
    if value == "regressed":
        return RegressedFailOn()

    match = _GEOMEAN_CONDITION_RE.fullmatch(value)
    if match is not None:
        return GeomeanFailOn(pct=float(match.group(1)))

    message = 'allowed values are "regressed" or "geomean:<number>" (e.g. geomean:2).'
    raise typer.BadParameter(message)


# ---------------------------------------------------------------------------
# Render mode
# ---------------------------------------------------------------------------


def resolve_render_mode(color_flag: bool) -> RenderMode:  # noqa: FBT001 -- 1:1 map of the --color flag
    """Map the ``--color`` flag to the output strategy the progress reporter uses.

    A vetoed flag suppresses color first; a non-TTY stderr always renders plain;
    a TTY renders a spinner when color is allowed and falls back to in-place
    overwrite otherwise.
    """
    if not color_flag:
        suppress_color()
    if not is_tty(sys.stderr):
        return "plain"
    color_allowed = color_flag and os.environ.get("NO_COLOR") is None
    return "spinner" if color_allowed else "overwrite"


# ---------------------------------------------------------------------------
# Repository lock
# ---------------------------------------------------------------------------


async def with_repo_lock[T](command: str, body: Callable[[], Awaitable[T]]) -> T:
    """Hold the repository's single-flight lock for the length of ``body``.

    Inside a git repository the lock is acquired around ``body`` and released
    however it settles — including on exception, and always before the caller
    renders its report. Outside every git repository the answer is to run
    ``body`` with no lock at all; any other git failure exits without
    benchmarking rather than running unlocked.
    """
    try:
        root = repo_root()
    except NotAGitRepositoryError:
        return await body()
    except GymratError as error:
        exit_with_error(error)

    release = acquire_lock(lockfile_path(root), command)
    try:
        return await body()
    finally:
        release()


# ---------------------------------------------------------------------------
# Flag dataclasses
# ---------------------------------------------------------------------------


@dataclass(frozen=True, slots=True)
class SharedFlags(CliFlags):
    """The flags every command carries: the config set plus how the report prints."""

    color: bool = True
    format: Literal["text", "json"] = "text"


@dataclass(frozen=True, slots=True)
class CompareFlags(SharedFlags):
    """The compare command's flags: the shared set plus the two only a verdict can answer."""

    verbose: bool = False
    fail_on: tuple[FailOnCondition, ...] = ()


@dataclass(frozen=True, slots=True)
class MeasureFlags(SharedFlags):
    """The measure command's flags: the shared set plus whether to record the run."""

    record: bool = False


# ---------------------------------------------------------------------------
# CLI option declarations
# ---------------------------------------------------------------------------

# The samples ceiling: JavaScript's ``Number.MAX_SAFE_INTEGER``, kept so the
# accepted range matches the shipped tool rather than drifting to a new bound.
_MAX_SAMPLES = 2**53 - 1


class OutputFormat(StrEnum):
    """The ``--format`` choices: a human report or a machine-readable document."""

    text = "text"
    json = "json"


# The config-bearing options every command shares, declared once as reusable
# annotations so ``compare`` and ``measure`` carry an identical surface.
BenchOption = Annotated[str | None, typer.Option("--bench", "-b", help="bench command")]
PrepareOption = Annotated[
    str | None,
    typer.Option("--prepare", "-p", help="preparation script to run before each revision"),
]
AdapterOption = Annotated[
    str | None, typer.Option("--adapter", "-a", help="adapter type for parsing benchmark output")
]
SamplesOption = Annotated[
    int | None,
    typer.Option(
        "--samples",
        "-s",
        parser=parse_positive_integer_up_to(_MAX_SAMPLES),
        help="paired samples per target",
    ),
]
TimeoutOption = Annotated[
    int | None,
    typer.Option(
        "--timeout",
        "-t",
        parser=parse_positive_integer_up_to(MAX_TIMEOUT_SECONDS),
        help="timeout in seconds",
    ),
]
ConfigOption = Annotated[str | None, typer.Option("--config", "-c", help="configuration file path")]
NoColorOption = Annotated[
    bool, typer.Option("--no-color", help="print the report without ANSI styles")
]
FormatOption = Annotated[OutputFormat, typer.Option("--format", help="output format")]
DebugOption = Annotated[bool, typer.Option("--debug", "-d", help="show stack traces on errors")]
RecordOption = Annotated[
    bool,
    typer.Option("--record", "-r", help="append the run to the session log as a baseline"),
]


# ---------------------------------------------------------------------------
# Run infrastructure
# ---------------------------------------------------------------------------


def begin_run(flags: SharedFlags, target_count: int) -> ProgressReporter:
    """Suppress color per the flag, then build the reporter for ``target_count`` targets."""
    mode = resolve_render_mode(flags.color)
    return create_progress_reporter(mode, target_count)


def run_options_of(config: ResolvedConfig, progress: ProgressReporter) -> RunOptions:
    """Wire ``config``'s run settings and ``progress``'s callbacks into the shared run fields."""
    return RunOptions(
        bench=config.bench,
        prepare=config.prepare,
        adapter=config.adapter,
        samples=config.samples,
        timeout_seconds=config.timeout_seconds,
        config_metrics=config.metrics,
        config_kinds=config.kinds,
        on_progress=progress.report,
        warn=progress.warn,
    )


@dataclass(frozen=True, slots=True)
class ReportRenderers[T]:
    """The text and JSON renderers a command hands ``emit_report`` for its result type."""

    text: Callable[[T, ReportOptions], str]
    json: Callable[[T], str]


def emit_report[T](
    result: T,
    flags: SharedFlags,
    renderers: ReportRenderers[T],
    render_opts: ReportOptions,
) -> None:
    """Render ``result`` per ``flags.format`` and write it to stdout.

    The text renderer honors ``render_opts`` — the ``--no-color`` veto rides in
    its ``color`` field, and the renderer defers to stdout's own TTY state when
    ``color`` is left unset. The JSON document is never styled, so it ignores
    ``render_opts`` entirely.
    """
    if flags.format == "json":
        output = renderers.json(result)
    else:
        output = renderers.text(result, render_opts)
    write_and_flush(sys.stdout, output + "\n")
